### ISSUE TITLE
Move CIDs to the "multiformats" project

### ISSUE THREAD

**Issue by Stebalien** [multiformats/cid#26]
So, when proposing to use CIDs as peer IDs, I got a bit of push-back as CIDs currently imply IPLD and bring in the concept of IPLD formats, links, dags, data-structures, etc. However, that doesn't *have* to be the case. Really, CIDs are just `(codec, content-addr)` tuples. That is, they're just "typed pointers".

Proposal: Move the CID spec/concept to multiformats.

This would mean that the *codec* in `<version><codec><mhash>` wouldn't *necessarily* map to a defined IPLD codec. Instead, it would just be *some* multicodec that may or may not have an IPLD format defined (yet).

Why?

* People can start using CIDs without thinking about IPLD. Typed pointers are *really* nice and it would be great to be able to use them everywhere without having to sell all of IPLD.
* We can retroactively define IPLD formats for these codecs. Then, when projects realize "oh shoot, I wish I had used IPLD", they'll have an easier time upgrading.

Why not?

* This may encourage users a proliferation of multicodecs/IPLD formats. In IPLD, we use multicodecs to indicate *structure* not semantics.

---

Thoughts? Should we just keep CIDs where they are but say "you can use these without the rest of IPLD" (not sure we can really sell that).

**Comment by mikeal** [comment]
Makes sense to me.

At some point, we're going to have to figure out a governance structure that is compatible with standards bodies for the multicodec registry. CID, and the constraints/concerns regarding codec proliferation, should be a part of that discussion.

**Comment by mikeal** [comment]
Also, the spec for CID has moved to `ipld/specs`, so we'll want to move that along with this repo.

**Comment by whyrusleeping** [comment]
Sounds good to me.

**Comment by daviddias** [comment]
👍 from me as well.

**Comment by mikeal** [comment]
I'm not admin in both of those orgs so I can't complete a migration :(

**Comment by Stebalien** [comment]
The migration will take a while anyways. I just wanted to see if there were any objections.

[note: @jbenet's fine with whatever increases adoption]

---

@tomaka

If CIDs didn't imply IPLD, would you feel better about using them as peer IDs? A CID is really just a `(codec, multihash)` tuple (see the issue summary). They're really on a layer below IPLD and don't actually *need* to drag in concepts like DAGs, formats, paths, etc.

Note: This won't really change anything from a technical perspective. The alternative is to just leave CIDs where they are (in IPLD) but say that they can be used independently. Moving them makes it clear that they can be used independent of IPLD, leaving them makes it clear that they're a building-block for IPLD.

**Comment by daviddias** [comment]
The JS people can start moving as we have it easier cause we don't use the github urls as import paths. The spec can be moved right away too

**Comment by daviddias** [comment]
Opened an issue on all the CID repos to point to this thread ^^

**Comment by ianopolous** [comment]
Sounds fine to me. Though you make it sound like the "rest of ipld" is large and complicated. In my opinion it is not. Or at least the minimum to use cids is not - cbor + a special cbor tag for merkle links.

**Comment by Stebalien** [comment]
> Though you make it sound like the "rest of ipld" is large and complicated. In my opinion it is not. 

It's not. However, it brings in a bunch of concepts (merkledags, formats, hash-linked data, etc.). Really, we might be better served by improving the documentation.

Basically, I don't want CIDs being a part of IPLD to scare people away from using them. I'd *rather* leave them here (absent additional motivation) but I'm highly biased towards "IPLD everywhere".

**Comment by mikeal** [comment]
I agree with @Stebalien. CID is an address and can be used in a lot of very simple cases that aren't even linked data.

Once you start using CID's **as links** (IPLD) you open the floodgates of knowledge and terms that are going to get thrown at you.

If you look at the main topics right now in the project it's mostly second order problems we're already facing working with these data-structures, like replication. Understanding all of that isn't necessary to derive value from CIDs.

**Comment by mikeal** [comment]
It sounds like we have a consensus.

I'm going to move this repo to multiformats and move the current spec from ipld/specs to this repo on Monday December 3rd if there are no objections.

**Comment by reardenlife** [comment]
I've got a question.  I am sorry if this is a wrong place to ask it, but I have to try.

I would like to use the same asymetric keypair both for Ethereum and IPFS in order to build a publication service. I.e. the user making an ETH transaction to the wallet of a service, the dApp scans the blockchain to find the transactions on its address and extracts the public key ... Then it has to convert that key to the address in IPNS.  This is where I stuck so far.

In the original IPFS whitepaper it was pointed out that NodeId is a hash of piblic key.  The problem is that crypto algos in Ethereum and IPFS are different.  I started to research the issue and found out about CIDv1 that should support multicodecs.  But was it ever implemented?

As related to the current thread - am I correctly undetstanding that NodeId is not even CIDv0, bit just a hash of RSA public key?

I mean, w/o tight coupling between cryptocurrency blockchains and IPFS ... the situation looks really grim.
